backend/apps/matching/configuration/properties.py

```diff
@@ -0,0 +1,244 @@
+from itertools import chain
+import string
+
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+import numpy as np
+
+
+class Property:
+
+    properties = None
+
+    def __init__(self, label=None, name=None, help_text=None, info_text=None, private=False):
+        """
+        Create a new property of a participant.
+
+        label: Label of the property that is shown in the UI
+        help_text: Help text that should be displayed on hover in the UI
+        info_text: Information that is visible next to the property in the UI
+        private: Is true if only site administrators should be able to view this property
+        """
+        if label is None:
+            raise ValueError("You need to set the value for label.")
+        self.label = _(label)  # the text that should appear when this property is referenced
+        self.name = name if name is not None else label.lower().replace(" ", "_")
+        if "--" in self.name:
+            raise ValueError(
+                f"The field name {self.name} is not allowed, since it contains '--'."
+                " You can explicitly chose a suitable column name by setting the 'name' kwarg."
+            )
+        if len(self.name) > 12:
+            raise ValueError(
+                f"The field name {self.name} is too long. You can use 12 characters at max."
+                f" If you did not set the 'name' before, please set in in the constructor. "
+            )
+        self.help_text = help_text
+        self.info_text = info_text
+        self.private = private
+
+    def get_model_field_names(self, prefix=None):
+        raise NotImplementedError
+
+    def get_model_fields(self):
+        raise NotImplementedError
+
+    def generate_random_assignment(self, rs=None):
+        raise NotImplementedError
+
+    def get_private_fields(self):
+        if self.private:
+            return [True for i in self.get_model_field_names()]
+        else:
+            if hasattr(self, "properties"):
+                return chain(*[p.get_private_fields() for p in self.properties])
```

Comment:
How can a property have properties? Is `properties` semantically correct? Is this the nicest design?

Reply:
its supposed to be a structure that can be iterated and group substructures like a tree. E.g. a `PropertyGroup` is grouping several properties. A conditional property has properties that are displayed only if a boolean property is selected.

I have not come up with a nicer design, and jonny as well